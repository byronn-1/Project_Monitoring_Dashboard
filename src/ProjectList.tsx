import { SHORT_STAGE_NAME_LIST } from "./data/ProjectListData";
import { useState } from "react";

import { ProjectTemplate } from "./data/StorageTypes";
import { useAppSelector } from "./app/store";
import { useDispatch } from 'react-redux'
import { toggleProjectComplete } from "./app/projectsSlice";



const ProjectList = () => {

  const [viewMoreArray, setViewMoreArray] = useState<boolean[]>([]);
  const [currentProjectId, setCurrentProjectId] = useState<number | undefined>();

  
  const changeViewMore = (id: number) => {
    if (viewMoreArray[id] === false) {
      setViewMoreArray((viewMoreArray) =>
        viewMoreArray.map((oldValue, i) => (i === id ? true : oldValue))
      );
    } else {
      setViewMoreArray((viewMoreArray) =>
        viewMoreArray.map((oldValue, i) => (i === id ? false : oldValue))
      );
    }
  };


  const dispatch = useDispatch();
  const projectListData: ProjectTemplate[]  = useAppSelector(state  => state.project.projects);

  
  const setStageComplete = (shortName: string) => {  
    dispatch(toggleProjectComplete({projectId: currentProjectId, shortName}))
  }

  const getProject = () => {
    return projectListData.find(project => currentProjectId === project.projectId)
  }

  const project = getProject();


  const getProjects = (project: ProjectTemplate) => {
    const result = [];

    if (viewMoreArray.length <= project.projectId) {
      setViewMoreArray((viewMoreArray) => [...viewMoreArray, true]);
    }

    result.push(
      <div key={`${project.projectId}`}>
        <div className="list-item-container">
          <div className="project-list-item">
            <h2 className="project-title">{`${project.name}`}</h2>
            <p className="initial-ideas">{`${project.projectDetails.initialIdeas}`}</p>
            <button
              className="update-project-btn"
              onClick={() => setCurrentProjectId(project.projectId)}
            >
              Update Project
            </button>
          </div>
          <button
            className="ViewMore-btn"
            aria-label="ViewMore-btn"
            value={`${project.projectId}`}
            onClick={() => changeViewMore(project.projectId)}
          >View More</button>
        </div>
        {!viewMoreArray[project.projectId] ? (
          <div key={`${project.projectId + project.name}`}className="more-info-card" style={{display: viewMoreArray[project.projectId] ? "none" : "flex"}}>
            <button>See More Details</button>
            <div className="project-stage-card">
              {SHORT_STAGE_NAME_LIST.map((projectStage, index) =>
              <div key={index}  style={{ backgroundColor: project.projectStages[projectStage].complete ? 'green' : 'red' }} className="project-stage-box">{projectStage}</div> )}
            </div>
            <a href="https://github.com/byronn-1/Project_Monitoring_Dashboard"> Source Code </a>
          </div>
        ) : (
          <div key={`${project.name}`}></div>
        )}
      </div>
    );
    return result;
  };

  return (
    <div className="project-list">
      <div className="scroll">
        {projectListData.map((project) => getProjects(project))}
      </div>
      {currentProjectId !== undefined && (
        <>
            <div className="modal">
              <h1>{project!.name}</h1>
              <div>
                {SHORT_STAGE_NAME_LIST.map((shortName) =>  <div className=".stage-selection-short" key={shortName} onClick={() => setStageComplete(shortName)}>{shortName}</div>)}
              </div>
              <button onClick={() => setCurrentProjectId(undefined)}>Save</button>
            </div>
            <div className="modal-overlay" />
        </>
      )}
    </div>
  );
};

export default ProjectList;
