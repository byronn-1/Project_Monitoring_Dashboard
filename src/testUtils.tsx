import React from "react";
import { Provider } from "react-redux";

import { createMemoryHistory } from "history";
import { unstable_HistoryRouter as HistoryRouter } from "react-router-dom";

import type { RenderOptions } from "@testing-library/react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { configureStore } from "@reduxjs/toolkit";
import type { PreloadedState } from "@reduxjs/toolkit";
import type { RootState } from "./app/store";
import projectReducer from './app/projectsSlice';


interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<Partial<RootState>>;
  initialRoutes?: string[];
}

// render with redux store and react router history
export function renderWithProvider(

  component: React.ReactElement,

  {
    preloadedState = undefined,
    initialRoutes = ["/"],
    ...renderOptions
  }: ExtendedRenderOptions = {}
)

{
  const store = configureStore({
    reducer: {project : projectReducer,  },
    preloadedState: {
      ...preloadedState,
    },

  });

  const user = userEvent.setup();
  let history = createMemoryHistory({ initialEntries: initialRoutes });
  
  return {
    store,
    user,
    history,
    ...render(
      <Provider store={store}>
        <HistoryRouter history={history}>{component}</HistoryRouter>
      </Provider>,
      renderOptions
    ),
  };
}

