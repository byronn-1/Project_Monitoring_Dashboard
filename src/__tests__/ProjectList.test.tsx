import { render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import ProjectList from "../ProjectList";
import { renderWithProvider } from "../testUtils";
import { Route, Routes } from "react-router-dom";


describe("Project list",  () => {

    it("renders correctly", () => {
        const { container } = renderWithProvider(<ProjectList />);
    
        expect(container).toMatchSnapshot();
    });
      
    it("Dropdown rendered when view more button clicked", async() => { 
        renderWithProvider(<ProjectList />);
        
        await userEvent.click(screen.getAllByText("View More")[3]);
        expect(screen.getByText("Brief")).toBeDefined();

    });

    it("update project modal rendered on button click", async () => {
        renderWithProvider(<ProjectList />);

        await userEvent.click(screen.getAllByText("Update Project")[0]);
        
        expect(screen.getByText("Save")).toBeDefined();

    });
    it("update view more dropdown icon colour on click modal click", async () => {
      renderWithProvider(<ProjectList />);

      await userEvent.click(screen.getAllByText("Update Project")[0]);
      await userEvent.click(screen.getByText("Brief"));
      await userEvent.click(screen.getByText("Save"));
      await userEvent.click(screen.getAllByText("View More")[0]);

      expect(screen.getByText("Brief")).toHaveStyleRule('backgroundColor', 'green');
  });

    function renderWithRoute(projectId: string) {
        // Add routes to get the contents of useParams populated
        return renderWithProvider(
          <Routes>
            <Route path="project/:projectId" element={<ProjectList />} />
            <Route path="*" element={<div>Unknown path</div>} />
          </Routes>,
          {
            initialRoutes: ["/", `/project/${projectId}`],
          }
        );
    }
});