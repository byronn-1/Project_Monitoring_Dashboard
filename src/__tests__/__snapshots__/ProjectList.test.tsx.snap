// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project list renders correctly 1`] = `
<div>
  <div
    class="project-list"
  >
    <div
      class="scroll"
    >
      <div>
        <div
          class="list-item-container"
        >
          <div
            class="project-list-item"
          >
            <h2
              class="project-title"
            >
              Inventory Project
            </h2>
            <p
              class="initial-ideas"
            >
              Application to keep track of projects amounts of products available and product details
            </p>
            <button
              class="update-project-btn"
            >
              Update Project
            </button>
          </div>
          <button
            aria-label="ViewMore-btn"
            class="ViewMore-btn"
            value="0"
          >
            View More
          </button>
        </div>
        <div />
      </div>
      <div>
        <div
          class="list-item-container"
        >
          <div
            class="project-list-item"
          >
            <h2
              class="project-title"
            >
              CSS React Project
            </h2>
            <p
              class="initial-ideas"
            >
              Experimental app to test creation of new components that can handle CSS
            </p>
            <button
              class="update-project-btn"
            >
              Update Project
            </button>
          </div>
          <button
            aria-label="ViewMore-btn"
            class="ViewMore-btn"
            value="1"
          >
            View More
          </button>
        </div>
        <div />
      </div>
      <div>
        <div
          class="list-item-container"
        >
          <div
            class="project-list-item"
          >
            <h2
              class="project-title"
            >
              Meal Planning Project
            </h2>
            <p
              class="initial-ideas"
            >
              Keep all favorite recipes in one  place
            </p>
            <button
              class="update-project-btn"
            >
              Update Project
            </button>
          </div>
          <button
            aria-label="ViewMore-btn"
            class="ViewMore-btn"
            value="2"
          >
            View More
          </button>
        </div>
        <div />
      </div>
      <div>
        <div
          class="list-item-container"
        >
          <div
            class="project-list-item"
          >
            <h2
              class="project-title"
            >
              Staffing Planner Project
            </h2>
            <p
              class="initial-ideas"
            >
              Create a calender app for staffing rota at multiple venues
            </p>
            <button
              class="update-project-btn"
            >
              Update Project
            </button>
          </div>
          <button
            aria-label="ViewMore-btn"
            class="ViewMore-btn"
            value="3"
          >
            View More
          </button>
        </div>
        <div />
      </div>
      <div>
        <div
          class="list-item-container"
        >
          <div
            class="project-list-item"
          >
            <h2
              class="project-title"
            >
              Web3 Project
            </h2>
            <p
              class="initial-ideas"
            >
              Create a calender app for staffing rota at multiple venues.
            </p>
            <button
              class="update-project-btn"
            >
              Update Project
            </button>
          </div>
          <button
            aria-label="ViewMore-btn"
            class="ViewMore-btn"
            value="4"
          >
            View More
          </button>
        </div>
        <div />
      </div>
      <div>
        <div
          class="list-item-container"
        >
          <div
            class="project-list-item"
          >
            <h2
              class="project-title"
            >
              Job Application Tracker
            </h2>
            <p
              class="initial-ideas"
            >
              App for tracking job applications the amount of and companies applied to
            </p>
            <button
              class="update-project-btn"
            >
              Update Project
            </button>
          </div>
          <button
            aria-label="ViewMore-btn"
            class="ViewMore-btn"
            value="5"
          >
            View More
          </button>
        </div>
        <div />
      </div>
    </div>
  </div>
</div>
`;
